api/grpcserver/v2alpha1/transaction.go

```diff
@@ -0,0 +1,463 @@
+)
+
+const (
+	Transaction       = "transaction_v2alpha1"
+	TransactionStream = "transaction_stream_v2alpha1"
+)
+
+// transactionConState is an API to validate transaction.
+type transactionConState interface {
+	Validation(raw types.RawTx) system.ValidationRequest
+}
+
+// transactionSyncer is an API to get sync status.
+type transactionSyncer interface {
+	IsSynced(context.Context) bool
+}
+
+// transactionValidator is the API to validate and cache transactions.
+type transactionValidator interface {
+	VerifyAndCacheTx(context.Context, []byte) error
+}
+
+func NewTransactionStreamService(db sql.Executor) *TransactionStreamService {
+	return &TransactionStreamService{db: db}
+}
+
+type TransactionStreamService struct {
+	db sql.Executor
+}
+
+func (s *TransactionStreamService) RegisterService(server *grpc.Server) {
+	spacemeshv2alpha1.RegisterTransactionStreamServiceServer(server, s)
+}
+
+func (s *TransactionStreamService) RegisterHandlerService(mux *runtime.ServeMux) error {
+	return spacemeshv2alpha1.RegisterTransactionStreamServiceHandlerServer(context.Background(), mux, s)
+}
+
+func (s *TransactionStreamService) Stream(
+	request *spacemeshv2alpha1.TransactionStreamRequest,
+	stream spacemeshv2alpha1.TransactionStreamService_StreamServer,
+) error {
+	return status.Errorf(codes.Unimplemented, "this endpoint has not yet been implemented")
+}
+
+func (s *TransactionStreamService) String() string {
+	return "TransactionStreamService"
+}
+
+func NewTransactionService(db sql.Executor, conState transactionConState,
+	syncer transactionSyncer, validator transactionValidator,
+	publisher pubsub.Publisher,
+) *TransactionService {
+	return &TransactionService{
+		db:        db,
+		conState:  conState,
+		syncer:    syncer,
+		validator: validator,
+		publisher: publisher,
+	}
+}
+
+type TransactionService struct {
+	db        sql.Executor
+	conState  transactionConState
+	syncer    transactionSyncer
+	validator transactionValidator
+	publisher pubsub.Publisher // P2P Swarm
+}
+
+func (s *TransactionService) RegisterService(server *grpc.Server) {
+	spacemeshv2alpha1.RegisterTransactionServiceServer(server, s)
+}
+
+func (s *TransactionService) RegisterHandlerService(mux *runtime.ServeMux) error {
+	return spacemeshv2alpha1.RegisterTransactionServiceHandlerServer(context.Background(), mux, s)
+}
+
+// String returns the service name.
+func (s *TransactionService) String() string {
+	return "TransactionService"
+}
+
+func (s *TransactionService) List(
+	ctx context.Context,
+	request *spacemeshv2alpha1.TransactionRequest,
+) (*spacemeshv2alpha1.TransactionList, error) {
+	ops, err := toTransactionOperations(request)
+	if err != nil {
+		return nil, status.Error(codes.InvalidArgument, err.Error())
+	}
+
+	switch {
+	case request.Limit > 100:
+		return nil, status.Error(codes.InvalidArgument, "limit is capped at 100")
+	case request.Limit == 0:
+		return nil, status.Error(codes.InvalidArgument, "limit must be set to <= 100")
+	}
```

Comment:
We should probably check this first to exit early before converting the request to operations.

Reply:
Changed order